### PR TITLE
Space - Liv

### PR DESCRIPTION
**Pull request by mulhoo**
# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | .top in Work, it returns the work with the most upvotes. 
Describe how you approached testing that model method. What edge cases did you come up with? | if there are no votes, it shouldn't return anything, if there is a tie, choose one. 
What are `session` and `flash`? What is the difference between them? | a session tracks the users movement until they log out; a flash is just a momentary thing that pops up to relay a message
What was one thing that you gained more clarity on through this assignment? | routes and schemas
What is the Heroku URL of your deployed application? | https://liv-media-ranks.herokuapp.com

### CODE REVIEW

File: app/models/work.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  validates :creator, presence: true
+
+  def self.top
+    return nil if Vote.count == 0
+
+    top_work = Work.all.max_by do |work|
+      work.votes.count
+    end
+
+    return top_work
```
Comment (dHelmgren) [inline app/models/work.rb:14]: This doesn't seem to work with your starting seeds. It doesn't return a  top media when one category is empty/there are no votes.

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Rails.application.routes.draw do
+  root 'homepage#index'
+  resources :works
+  resources :users
```
Comment (dHelmgren) [inline config/routes.rb:4]: Users doesn't need all of the routes that `resources:` provides. you should aim to limit the number of routes in a project!

File: app/views/homepage/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+          </div>
+        </ol>
+      <% end %>
+    </ul>
+  </section>
+
+   <section class="top-ten_list-container">
+    <h3 class="top-ten__header">Top Albums</h3>
+    <ul class='list-group top-ten__list'>
+      <% @sort_albums[0..9].each do |album| %>
```
Comment (dHelmgren) [inline app/views/homepage/index.html.erb:55]: You repeat this HTML 3 times, making it a good candidate for a view partial. They aren't just for forms!

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+class User < ApplicationRecord
+  has_many :votes
```
Comment (dHelmgren) [inline app/models/user.rb:2]: This is where you can put the restriction for limiting how many times a user can vote on works.

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,89 @@
+class UsersController < ApplicationController
+  before_action :require_login, only: [:show]
```
Comment (dHelmgren) [inline app/controllers/users_controller.rb:2]: NICE

**Comment by dHelmgren** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️ (see comment)
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | no details about the failed save!
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) |
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | no votes table in works show page
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | I can over vote
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | still generic

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | see comment
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate |
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works |
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review |
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | ✔️
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |
